contracts/lib.rs

```diff
@@ -469,258 +462,257 @@ pub mod prosopo {
 
-            // provider cannot be the admin
-            self.check_not_admin(provider_account)?;
+            self.service_origins.remove(old_service_origin_hash);
+            // don't record the default hash of the service origin as this is a special placeholder hash which is used elsewhere, e.g. in testing / setting up a dummy or default provider, so multiple providers may have this hash set
+            if new_service_origin_hash != default_dataset_id {
+                self.service_origins.insert(new_service_origin_hash, &());
+            }
 
-            let service_origin_hash = self.hash_vec_u8(&service_origin);
+            self.providers.insert(provider_account, &new_provider);
 
-            // prevent duplicate service origins
-            if self.service_origins.get(service_origin_hash).is_some() {
-                return err!(Error::ProviderServiceOriginUsed);
+            // update the category if status or payee has changed
+            if old_provider.status != new_provider.status
+                || old_provider.payee != new_provider.payee
+            {
+                self.provider_state_remove(&old_provider)?;
+                self.provider_state_insert(&new_provider)?;
             }
-            let balance: Balance = 0;
-            // add a new provider
-            let provider = Provider {
-                status: GovernanceStatus::Deactivated,
-                balance,
-                fee,
-                service_origin,
-                dataset_id: Hash::default(),
-                payee,
-            };
-            self.providers.insert(provider_account, &provider);
-            self.service_origins.insert(service_origin_hash, &());
-            let mut provider_accounts_map = self
-                .provider_accounts
-                .get(ProviderState {
-                    status: GovernanceStatus::Deactivated,
-                    payee,
-                })
-                .unwrap_or_default();
-            provider_accounts_map.insert(provider_account);
-            self.provider_accounts.insert(
-                ProviderState {
-                    status: GovernanceStatus::Deactivated,
-                    payee,
-                },
-                &provider_accounts_map,
-            );
+
             Ok(())
         }
 
-        /// Update an existing provider, their service origin, fee and deposit funds
-        #[ink(message)]
-        #[ink(payable)]
-        pub fn provider_update(
-            &mut self,
-            service_origin: Vec<u8>,
-            fee: u32,
-            payee: Payee,
-        ) -> Result<(), Error> {
+        /// Remove the provider from their state
+        fn provider_state_remove(&mut self, provider: &Provider) -> Result<(), Error> {
             let provider_account = self.env().caller();
 
-            // this function is for updating only, not registering
-            if self.providers.get(provider_account).is_none() {
+            let cat = ProviderState {
```

Comment:
Please use `category`. We may have non-English speaking developers at some point and abbreviations could be confusing.